openwisp-monitoring/files/monitoring.agent

```diff
@@ -276,16 +278,34 @@ main() {
 	INTERVAL=${INTERVAL:-300}
 	REGISTRATION_INTERVAL=$((INTERVAL / 10))
 	VERBOSE_MODE=${VERBOSE_MODE:-0}
+	BOOTUP_DELAY=${BOOTUP_DELAY:-10}
 	TMP_DIR="/tmp/openwisp/monitoring"
+	BOOTUP="$TMP_DIR/bootup"
 
 	[ -z "$MODE" ] && echoerr "missing required --mode option"
 
+	# make directory
+	mkdir -p "$TMP_DIR"
+
 	if [ "$MODE" = "collect" ]; then
 		MONITORED_INTERFACES=${MONITORED_INTERFACES:-*}
 		# remove double quotes from interfaces
 		MONITORED_INTERFACES=$(echo "$MONITORED_INTERFACES" | tr -d '"')
 		save_data
 	elif [ "$MODE" = "send" ]; then
+		if [ ! -f "$BOOTUP" ]; then
+			# actions to do only after agent start, not after restart
+			if [ "$BOOTUP_DELAY" -ne "0" ]; then
+				# get a random number between zero and $BOOTUP_DELAY
+				DELAY=$(/usr/sbin/openwisp-get-random-number 0 "$BOOTUP_DELAY")
+				logger "Delaying sending metrics for $DELAY seconds..." \
+					-t openwisp-monitoring \
+					-p daemon.info
+				sleep "$DELAY"
+			fi
+			# send bootup hotplug event
+			touch "$BOOTUP"
+		fi
```

Comment:
While testing these changes locally I found that this part of code is executed twice 

```text
root@02-tplink-archer:~# rm -rf /tmp/openwisp/monitoring/;  /etc/init.d/openwisp-monitoring restart; logread -f
openwisp-monitoring: OpenWISP monitoring daemon stopping
openwisp-monitoring: OpenWISP monitoring daemon started
Fri Dec  9 16:50:26 2022 daemon.info openwisp: Delaying startup for 5 seconds...
Fri Dec  9 16:50:32 2022 daemon.info openwisp: Delaying startup for 4 seconds...
```

Upon deeper investigation, I found that the script exists unexpectedly while executing the `sleep` statement (i.e. the statements after sleep are not executed). When the script executes for the second time, the statements following `sleep` are executed. When I comment out the `sleep` statement the script runs without any issues (no duplicate delays). 

I wonder if there is something that kills the script when it executes the `sleep` statement. I didn't find anything fruitful while running the openwisp-monitoring package in verbose mode. I also tried restarting the openwisp-monitoring agent with `PROCD_DEBUG=1` as mentioned in https://openwrt.org/docs/guide-developer/procd-init-scripts#debugging

Reply:
Thanks to @nemesisdesign, I finally found the reason why the script was getting reloaded. It was due to the `collect_data` function sending `SIGUSR1` signal. 

https://github.com/openwisp/openwrt-openwisp-monitoring/blob/6d9340d62cc502d1a1e85fdbb77460c836004d98/openwisp-monitoring/files/monitoring.agent#L109-L112